README.md
[![ci-cd](https://github.com/UBC-MDS/pokehelpyer/actions/workflows/ci-cd.yml/badge.svg)](https://github.com/UBC-MDS/pokehelpyer/actions/workflows/ci-cd.yml)

# pokehelpyer

`pokehelpyer` is a Python package designed to assist Pokémon players in building teams of pokémon. Users can provide a list of pokémon currently on their team, and `pokehelpyer` will recommend a suitable pokémon to add to the team based on the current team's overall weaknesses and resistances.

In the Pokémon game, players battle using teams of up to six pokémon. Each pokémon has a type, such as grass, water, or fire. Each type in the game is either weak to, resistant to, or neutrally affected by each other type in the game. If a player does not consider these weaknesses and resistances when creating their team of pokémon, they may find themselves in a scenario in which they are consistently struggling to win battles against certain types of pokémon. For a simple example, if a player were to use a team of six fire-type pokémon, then they would likely struggle to win battles against water-type pokémon. The preceding is obviously an extreme example, but in a competitive setting, even a small imbalance in a team's weaknesses and resistances can cause major issues for a player.

The `pokehelpyer` package aims to assist players in avoiding imbalances in weaknesses/resistances when building a pokémon team. Given a list of up to five pokémon currently on a team, `pokehelpyer` will determine which types the team is most weak to and which types the team is most resistant to. Based on this information, `pokehelpyer` will then recommend a pokémon that could be added to the current team to make its weaknesses and resistances more balanced.

(If you are reading this and aren't familiar with Pokémon, it might be helpful to note that "pokémon" is both the singular and plural form of the word, and also that lowercase "pokémon" refers to an individual character in the game, such as Pikachu, whereas capitalized "Pokémon" refers to the franchise/game.)

## Functions Included

- `get_types`: Given a list of pokémon names, determine the types of those pokémon using an existing dataset.
<br>

- `calc_resistances`:  Given a list of pokémon types present on a player's team,
    calculate a measure of how resistant the team is to each type in the game.

  - Creates a dictionary in which the keys are each of the 18 types
    in the game, and the values are integers measuring the level of
    resistance the input team has to that key (type). Higher values indicate a
    higher level of resistance to that type.
<br><br>

- `calc_weaknesses`: Given a list of pokémon types present on a player's team,
    calculate a measure of how weak the team is to each type in the game.

  - Creates a dictionary in which the keys are each of the 18 types
    in the game, and the values are integers measuring the level of
    weakness the input team has to that key (type). Higher values indicate a
    higher level of weakness to that type.
<br><br>

- `recommend`: Given a team of up to 5 pokémon, recommend a
    pokémon that could be added to the
    current team to make its weaknesses and
    resistances more balanced.
  - Determines which types the
    team is most resistant to and weak to via `calc_resistances` and
    `calc_weaknesses`, and then makes its recommendation
    based on this information.
  - In particular, it uses `calc_balance` together a brute-force search of
    all ~700 pokémon to determine its recommendation based on the objective of
    maximizing balance.
<br><br>

- `calc_balance`: Calculate a measure of how balanced a pokémon team is using its
    weaknesses and resistances. Higher values indicate a more balanced team.

## Datasets Included

- [pokémon.csv](https://github.com/UBC-MDS/pokehelpyer/blob/main/data/pokemon.csv): Dataset containing details about each pokémon including its name, type(s), and base stats.
- [type_chart.csv](https://github.com/UBC-MDS/pokehelpyer/blob/main/data/type_chart.csv): Dataset containing details abouts the weaknesses and strengths of each type of pokémon.

## Place Within the Python Ecosystem

There are websites and applications that help build pokémon teams, such as the [Mariland Team Builder](https://marriland.com/tools/team-builder/en/). However these tools simply present the player with a visual representation of their current team's weaknesses and resistances. They don't make recommendations. In other words, the existing tools simply given visual representations of the dictionaries created by `calc_weaknesses` and `calc_resistances`. There doesn't seem to be any existing Python packages which will use the weaknesses/resistances data to make reccomendations for additional team members.

## Installation

```bash
pip install pokehelpyer
```

## Usage

See https://pokehelpyer.readthedocs.io/en/latest/

```python
import pokehelpyer.pokehelpyer as pk
```

<br>

`get_types`:

```python
team_list = ['Pikachu', 'Eevee', 'Charizard', 'Metapod']
team_types = pk.get_types(team_list)
```

<br>

`calc_weaknesses` and `calc_resistances`:

```python
team_weaknesses = pk.calc_weaknesses(team_types)
team_resistances = pk.calc_resistances(team_types)
```

<br>

`calc_balance`:

```python
team_balance = pk.calc_balance(
    resistances=team_resistances, 
    weaknesses=team_weaknesses
)
```

`recommend`:

```python
pk.recommend(['Pikachu', 'Eevee', 'Charizard', 'Metapod'], n_recommendations=2)
```

## Contributors

- Raul Aguilar
- Jakob Thoms
- Ritisha Sharma
- Sneha Sunil

## Contributing

Interested in contributing? Check out the [contributing guidelines](https://github.com/UBC-MDS/pokehelpyer/blob/main/CONTRIBUTING.md). Please note that this project is released with a [Code of Conduct](https://github.com/UBC-MDS/pokehelpyer/blob/main/CONDUCT.md). By contributing to this project, you agree to abide by its terms.

## Credits

`pokehelpyer` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).

## License

`pokehelpyer` was created by Jakob Thoms, Ritisha Sharma, Raul Aguilar, and Sneha Sunil. It is licensed under the terms of the MIT license.

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)  


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.8 (2023-02-05)


## v0.1.7 (2023-01-29)


## v0.1.6 (2023-01-29)


## v0.1.5 (2023-01-29)


## v0.1.4 (2023-01-29)


## v0.1.3 (2023-01-29)


## v0.1.2 (2023-01-29)


## v0.1.1 (2023-01-29)


## v0.1.0 (2023-01-29)
### Feature
* Add input checks to calc_balance ([`9c67136`](https://github.com/UBC-MDS/pokehelpyer/commit/9c67136e69745b5d600b0ecdd299560fbaacc089))
* Add optional arguments to the recommend function ([`6891d0e`](https://github.com/UBC-MDS/pokehelpyer/commit/6891d0e089fda348469dc98ccab7999e7194fe95))
* Code the calc_balance function ([`ef7e948`](https://github.com/UBC-MDS/pokehelpyer/commit/ef7e9487b42642b028f49610353e6efc43563e33))
* Code the recommend function ([`b8d3e81`](https://github.com/UBC-MDS/pokehelpyer/commit/b8d3e81a47e095b123202bd6c68ac9c185f00cc0))
* Code part of the recommend function ([`7493be2`](https://github.com/UBC-MDS/pokehelpyer/commit/7493be205d5b4508f545b07270ba2ca65a5964bd))
* Add skeleton for calc_balance function ([`ed82989`](https://github.com/UBC-MDS/pokehelpyer/commit/ed82989cf7dce820107f082db52892d43581b1fc))

### Documentation
* Update package version ([`4396ab0`](https://github.com/UBC-MDS/pokehelpyer/commit/4396ab0513c6385a811f46bbf154ccff125d89d1))
* Add examples to calc_balance docstring ([`69e2a3b`](https://github.com/UBC-MDS/pokehelpyer/commit/69e2a3b4ed7dc92ce4df9fba97ee1b16a7d89942))
* Small update to recommend docstring ([`8dbda3c`](https://github.com/UBC-MDS/pokehelpyer/commit/8dbda3c229534fdef7ab43f6a2ac3feb5f72454b))
* Revise docstring of calc_weaknesses ([`a20cacb`](https://github.com/UBC-MDS/pokehelpyer/commit/a20cacb24a3e744cc30784c84081e91649676ef7))
* Revise docstring of calc_resistances ([`18db0ec`](https://github.com/UBC-MDS/pokehelpyer/commit/18db0ecf6029cda332644e32dac5272e7e4d6e67))
* Add docstring and skeleton for recommend function ([`b68cc47`](https://github.com/UBC-MDS/pokehelpyer/commit/b68cc47cc3349f0f566ebb60e4e236f54231becd))
* Minor edits to function documentations ([`3e1d099`](https://github.com/UBC-MDS/pokehelpyer/commit/3e1d099b52d703846b67e8c87d0a5fa3b7722d3d))
* Fix spelling of pokémon in readme (add accent to é) ([`d481111`](https://github.com/UBC-MDS/pokehelpyer/commit/d481111f337c83466d5243a94586bdda8a57785d))
* Edit README.md ([`406ff45`](https://github.com/UBC-MDS/pokehelpyer/commit/406ff4581c29d1a2e99e6403c9524b18ab8f0a3f))
* Edit CONTRIBUTING.md ([`44f9160`](https://github.com/UBC-MDS/pokehelpyer/commit/44f91609fab3279179e43a10761dea6fe9cf567f))
* Add github links to CONTRIBUTORS.md ([`369e0bc`](https://github.com/UBC-MDS/pokehelpyer/commit/369e0bcf1bc7057129fa3d70dd1f8ae0cd0ba930))
* Add CONTRRIBUTORS.md ([`151aac5`](https://github.com/UBC-MDS/pokehelpyer/commit/151aac5defff54808f671607988b246f1e1eaa4c))

## v0.1.0 (10/01/2023)

- First release of `pokehelpyer`!

tests/test_pokehelpyer.py
from pokehelpyer.pokehelpyer import (
    get_types,
    calc_weaknesses,
    calc_resistances,
    recommend,
    calc_balance,
)


# get_types:


def test_get_types():
    """Test get_types function returns correct result for different inputs."""
    assert get_types(["RaICHu", "zubAt"]) == [
        ["Electric"],
        ["Poison", "Flying"],
    ], "Not able to handle improper capitalization"
    assert get_types(["Snorlax!!", "...Chingling"]) == [
        ["Normal"],
        ["Psychic"],
    ], "Not able to handle improper punctuation"
    assert get_types(["Mime Jr.", "Porygon-Z", "Nidoran♀"]) == [
        ["Psychic", "Fairy"],
        ["Normal"],
        ["Poison"],
    ], "Not able to handle Pokemon with symbol in the name"


# calc_weaknesses:


def test_calc_weaknesses_valid_result():
    """Tests the calculate weaknesses function for
        valid output/positive testcase"""
    input_list = [["Fire"], ["Steel", "Flying"], ["Grass", "Ice"]]
    actual_dict = calc_weaknesses(input_list)
    expected_dict = {
        "Normal": 0,
        "Fire": 3,
        "Water": 1,
        "Electric": 1,
        "Grass": 0,
        "Ice": 0,
        "Fighting": 1,
        "Poison": 1,
        "Ground": 1,
        "Flying": 1,
        "Psychic": 0,
        "Bug": 1,
        "Rock": 2,
        "Ghost": 0,
        "Dragon": 0,
        "Dark": 0,
        "Steel": 1,
        "Fairy": 0,
    }
    assert actual_dict == expected_dict
    assert isinstance(actual_dict, dict)


def test_calc_weaknesses_invalid_result():
    """Tests the calculate weaknesses function for negative testcase"""
    actual_dict = calc_weaknesses([["Steel", "Flying"], ["Grass", "Ice"]])
    expected_dict = {
        "Normal": 0,
        "Fire": 0,
        "Water": 1,
        "Electric": 1,
        "Grass": 0,
        "Ice": 0,
        "Fighting": 1,
        "Poison": 1,
        "Ground": 1,
        "Flying": 1,
        "Psychic": 0,
        "Bug": 1,
        "Rock": 2,
        "Ghost": 0,
        "Dragon": 0,
        "Dark": 0,
        "Steel": 1,
        "Fairy": 0,
    }
    assert actual_dict != expected_dict


def test_calc_weaknesses_invalid_input():
    """Tests the calculate weaknesses function for empty inputs"""
    actual_val = calc_weaknesses([])
    assert actual_val is None
    actual_val = calc_weaknesses([[], []])
    assert actual_val is None


# calc_resistances:


def test_calc_resistances():
    """Test calc_resistances functionality."""
    assert (
        sum(calc_resistances([["Normal"]]).values()) == 3
    ), "Pokemon type does not add 3 points for immunity"
    assert (
        calc_resistances([["Steel", "Flying"]])["Grass"] == 2
    ), "Pokemon type does not add 2 points for double resistance"
    assert (
        sum(calc_resistances([["Dark", "Fairy"]]).values())
        == 3 + 3 + 2 + 1
    ), """Combination of immunity, double resistance or
        single resistance not being calculated properly"""


# recommend:


def test_recommend():
    """Test `recommend` returns the correct type of output."""
    actual = recommend(
        ["Pikachu", "Charizard"], n_recommendations=1, early_stop=False
    )  # Do a full run for one test. Early stop for the rest.
    assert isinstance(
        actual, str
    ), "recommend with n_recommendations=1 should return a string."

    n = 3
    actual = recommend(
        ["Pikachu", "Charizard"], n_recommendations=n, early_stop=True
    )
    assert isinstance(
        actual, list
    ), "recommend with n_recommendations > 1 should return a list of strings."
    assert isinstance(
        actual[0], str
    ), "recommend with n_recommendations > 1 should return a list of strings."
    assert (
        len(actual) == n
    ), """recommend with n_recommendations > 1 should return a
        list of strings with length equal to n_recommendations."""

    """Test `recommend` deals with errorneous user input
        in the case of a team with only one pokémon."""
    actual = recommend("Pikachu", early_stop=True)
    expected = recommend(["Pikachu"], early_stop=True)
    assert actual == expected


# calc_balance:


def test_calc_balance():
    zero_resistances = {
        "Normal": 0,
        "Fire": 0,
        "Water": 0,
        "Electric": 0,
        "Grass": 0,
        "Ice": 0,
        "Fighting": 0,
        "Poison": 0,
        "Ground": 0,
        "Flying": 0,
        "Psychic": 0,
        "Bug": 0,
        "Rock": 0,
        "Ghost": 0,
        "Dragon": 0,
        "Dark": 0,
        "Steel": 0,
        "Fairy": 0,
    }
    zero_weaknesses = zero_resistances
    fire_water_resistances = {
        "Normal": 0,
        "Fire": 1,
        "Water": 1,
        "Electric": 0,
        "Grass": 0,
        "Ice": 0,
        "Fighting": 0,
        "Poison": 0,
        "Ground": 0,
        "Flying": 0,
        "Psychic": 0,
        "Bug": 0,
        "Rock": 0,
        "Ghost": 0,
        "Dragon": 0,
        "Dark": 0,
        "Steel": 0,
        "Fairy": 0,
    }
    fire_water_weaknesses = fire_water_resistances
    all_resistances = {
        "Normal": 1,
        "Fire": 1,
        "Water": 1,
        "Electric": 1,
        "Grass": 1,
        "Ice": 1,
        "Fighting": 1,
        "Poison": 1,
        "Ground": 1,
        "Flying": 1,
        "Psychic": 1,
        "Bug": 1,
        "Rock": 1,
        "Ghost": 1,
        "Dragon": 1,
        "Dark": 1,
        "Steel": 1,
        "Fairy": 1,
    }
    all_weaknesses = all_resistances

    """Test `calc_balance` returns a float."""
    assert isinstance(
        calc_balance(
            zero_resistances,
            zero_weaknesses,
        ),
        float,
    ), "calc_balance should return a float."
    assert isinstance(
        calc_balance(fire_water_resistances, fire_water_weaknesses), float
    ), "calc_balance should return a float."

    """Test `calc_balance` returns reasonable values."""
    assert (
        calc_balance(zero_resistances, zero_weaknesses) == 0
    ), """calc_balance should return zero for a team
        with no resistances nor weaknesses."""
    assert (
        calc_balance(fire_water_resistances, fire_water_weaknesses) == 0
    ), """calc_balance should return zero for a team
        with equal and opposite resistances and weaknesses."""
    assert (
        calc_balance(fire_water_resistances, zero_weaknesses) > 0
    ), """calc_balance should return a positive number for a team
        with resistances and no weaknesses."""
    assert calc_balance(all_resistances, zero_weaknesses) > calc_balance(
        fire_water_resistances, zero_weaknesses
    ), """calc_balance should return higher numbers
        for teams with more resistances."""
    assert (
        calc_balance(zero_resistances, fire_water_weaknesses) < 0
    ), """calc_balance should return a negative number
        for a team with weaknesses and no resistances."""
    assert calc_balance(zero_resistances, all_weaknesses) < calc_balance(
        zero_resistances, fire_water_weaknesses
    ), """calc_balance should return lower numbers
        for teams with more weaknesses."""


plot/type_advantage_plot.R
library(tidyverse)
library(latex2exp)

alpha <- 3/4
beta <- 3/2

dat = tibble(
  delta = seq(-8, 8),
  y_pos = delta^alpha,
  y_neg = -(-delta)^beta
)
head(dat, 20)

type_advantage_plot <- ggplot(dat, aes(x = delta)) + 
  geom_point(aes(y = y_neg), color = 'blue', size = 2.5) +
  geom_line(aes(y = y_neg), lwd = 0.8) +
  geom_point(aes(y = y_pos), color = 'red', size = 2.5) +
  geom_line(aes(y = y_pos), lwd = 0.8) +
  geom_vline(xintercept = 0) +
  geom_hline(yintercept = 0) +
  scale_x_continuous(n.breaks = 18) +
  scale_y_continuous(n.breaks = 18) +
  ggtitle("Type-Advantage Function") +
  labs(
    x = TeX(r"($ \overset{\delta_{type}}{(\phantom{.}_{=\ resistance_{type}\ -\ weakness_{type}\ })} $)"),
    y = "Type-Advantage"
  ) +
  theme_bw(base_size = 14) +
  theme(
    plot.title = element_text(size = 16, face = "bold"),
    axis.title = element_text(size = 15),
    axis.text.x = element_text(size = 13),
    axis.text.y = element_text(size = 13)
  )

type_advantage_plot



src/pokehelpyer/pokehelpyer.py
import pandas as pd
import re


def get_types(pokemon_names):
    """
    Given a list of pokémon names, determine the types of
    those pokémon using an existing dataset.

    Parameters
    ----------
    pokemon_names : list of strings
        list of pokémon names

    Returns
    -------
    pokemon_types : list of lists of strings
        list of lists of pokémon types corresponding to
        the pokémon names in the input list

    Example
    --------
    >>> get_types(['Pikachu', 'Eevee', 'Charizard', ...])
    [['Electric'], ['Normal'], ['Fire', 'Flying'], ...]
    """
    # Handle user input error in case of single pokemon
    if isinstance(pokemon_names, str):
        pokemon_names = [pokemon_names]
    # Check user input
    try:
        assert isinstance(
            pokemon_names, list
        ), f'{"Input should be a list of pokemon names."}'
        assert (
            len(pokemon_names) > 0
        ), f'{"Input should be a non-empty list of pokemon names."}'
        assert isinstance(
            pokemon_names[0], str
        ), f'{"Input should be a list of pokemon names."}'
    except AssertionError as ex:
        print(f"Invalid input: {ex}")
        return None

    # Read file with Pokemon names and types,
    # and keep only relevant subset of the data
    url = (
        "https://raw.githubusercontent.com/"
        + "UBC-MDS/pokehelpyer/main/data/pokemon.csv"
    )

    names_types_df = pd.read_csv(url)[["Name", "Type 1", "Type 2"]]

    # Clean string column "Name"
    names_types_df["Name"] = names_types_df["Name"].str.strip()
    names_types_df["Name"] = names_types_df["Name"].str.lower()
    # Some of the names in dataset contain symbols
    name_with_symbol = names_types_df[
        names_types_df["Name"].str.match(r".*[^\w\s].*")
    ]["Name"].tolist()

    poke_types = []
    for name in pokemon_names:
        # Clean string
        name = name.strip()
        name = name.lower()
        # Check if name is supposed to contain symbol, otherwise remove
        if name not in name_with_symbol:
            name = re.sub(r"[^\w\s]", "", name)

        # Check if exists in data
        assert (
            name in names_types_df["Name"].tolist()
        ), f'{"{name} is not a valid pokémon."}'

        # Find the row with match
        row = names_types_df.loc[names_types_df["Name"] == name].values[0]
        poke_types.append([row[1], row[2]])
        if pd.isna(row[2]):  # Type 2 is optional
            poke_types[-1].pop()

    return poke_types


def calc_resistances(team_types):
    """
    Given a list of pokémon types present on a player's team,
    calculate a measure of how resistant the team is to each type in the game.

    Creates a dictionary in which the keys are each of the 18 types
    in the game, and the values are integers measuring the level of
    resistance the input team has to that type. Higher values indicate a
    higher level of resistance to that type (key).

    Parameters
    ----------
    team_types : list of list of strings
        list of pokémon types associated to the user's team
        obtained via `get_types`

    Returns
    -------
    resistances : dictionary
        a dictionary containing all 18 pokémon types as keys,
        and integers measuring the level of resistance the input team
        has to that type as values.

    Examples
    --------
    >>> calc_resistances([['Electric'], ['Normal'],
                          ['Fire', 'Flying']])
    {'Normal': 0, 'Fire': 1, 'Water': 0, 'Grass': 2, 'Electric': 1, ...}

    >>> calc_resistances([['Steel', 'Flying']])
        # Skarmory is doubly resistant to Grass
    {'Normal': 1, 'Fire': 0, 'Water': 0, 'Grass': 2, 'Electric': 0, ...}

    """
    # Check user input
    try:
        assert isinstance(
            team_types, list
        ), f'{"Input should be a list of lists of pokemon types."}'
        assert (
            len(team_types) > 0
        ), f'{"Input should be a non-empty list of lists of pokemon types."}'
        assert isinstance(
            team_types[0], list
        ), f'{"Input should be a list of lists pokemon types."}'
        assert (
            len(team_types[0]) > 0
        ), f'{"Input should be a list of non-empty lists of pokemon types."}'
        assert isinstance(
            team_types[0][0], str
        ), f'{"Input should be a list of lists pokemon types."}'
    except AssertionError as ex:
        print(f"Invalid input: {ex}")
        return None

    # Loads the file with Pokemon types,
    # strengths and weaknesses into a data frame
    url = (
        "https://raw.githubusercontent.com/"
        + "UBC-MDS/pokehelpyer/main/data/type_chart.csv"
    )
    resistances_df = pd.read_csv(url, index_col=0)

    # Creating dictionary with all Pokemon types
    all_types = resistances_df.index.tolist()
    resistances = {x: 0 for x in all_types}

    for attacking_type in all_types:
        for type_combo in team_types:
            val1 = resistances_df.loc[attacking_type][type_combo[0]]
            if len(type_combo) == 1:
                val2 = 1
            else:
                val2 = resistances_df.loc[attacking_type][type_combo[1]]

            if val1 == 0 or val2 == 0:
                resistances[attacking_type] += 3
            elif (val1 == 0.5 and val2 == 2) or (
                val1 == 2 and val2 == 0.5
            ):
                continue
            elif val1 == 0.5 and val2 == 0.5:
                resistances[attacking_type] += 2
            elif (val1 == 1 and val2 == 0.5) or (
                val1 == 0.5 and val2 == 1
            ):
                resistances[attacking_type] += 1

    return resistances


def calc_weaknesses(team_types):
    """
    Given a list of pokémon types present on a player's team,
    calculate a measure of how weak the team is to each type in the game.

    Creates a dictionary in which the keys are each of the 18 types
    in the game, and the values are integers measuring the level of
    weakness the input team has to that key (type). Higher values indicate a
    higher level of weakness to that type.

    Parameters
    ----------
    team_types : list of list of strings
        list of pokémon types associated to the user's team
        obtained via `get_types`

    Returns
    -------
    weaknesses : dictionary
        a dictionary containing all 18 pokémon types as keys,
        and integers measuring the level of weakness the input team
        has to that type as values.

    Examples
    --------
    >>> calc_weaknesses([['Electric'], ['Normal'], ['Fire', 'Flying']])
    {'Normal': 0, 'Fire': 0, 'Water': 1, 'Grass': 0, 'Electric': 1, ...}

    >>> calc_weaknesses([['Ice', 'Grass']) # Abomasnow is doubly weak to Fire
    {'Normal': 0, 'Fire': 2, 'Water': 0, 'Grass': 0, 'Electric': 0, ...}

    """
    # Check user input
    try:
        assert isinstance(
            team_types, list
        ), f'{"Input should be a list of lists of pokemon types."}'
        assert (
            len(team_types) > 0
        ), f'{"Input should be a non-empty list of lists of pokemon types."}'
        assert isinstance(
            team_types[0], list
        ), f'{"Input should be a list of lists pokemon types."}'
        assert (
            len(team_types[0]) > 0
        ), f'{"Input should be list of non-empty lists of pokemon types."}'
        assert isinstance(
            team_types[0][0], str
        ), f'{"Input should be a list of lists pokemon types."}'
    except AssertionError as ex:
        print(f"Invalid input: {ex}")
        return None

    # Read the pokemon weakness dataframe using pandas
    url = (
        "https://raw.githubusercontent.com/"
        + "UBC-MDS/pokehelpyer/main/data/type_chart.csv"
    )
    weakness_df = pd.read_csv(url, index_col=0)

    # Fetch all types of pokemon
    all_types = weakness_df.index.tolist()
    weaknesses = {x: 0 for x in all_types}

    # Calculate weaknesses of all types and add it to a dictionary
    for attacking_type in all_types:
        for type_combo in team_types:
            val1 = weakness_df.loc[attacking_type][type_combo[0]]
            if len(type_combo) == 1:
                val2 = 1
            else:
                val2 = weakness_df.loc[attacking_type][type_combo[1]]

            if val1 == 0 or val2 == 0:
                continue
            elif (val1 == 0.5 and val2 == 2) or (
                val1 == 2 and val2 == 0.5
            ):
                continue
            elif val1 == 2 and val2 == 2:
                weaknesses[attacking_type] += 2
            elif (val1 == 1 and val2 == 2) or (val1 == 2 and val2 == 1):
                weaknesses[attacking_type] += 1

    return weaknesses


def recommend(
    current_team,
    n_recommendations=1,
    include_legendaries=False,
    include_megas=False,
    verbose=True,
    early_stop=False,
):
    """
    Given a team of up to 5 pokémon, recommend a
    pokémon that could be added to the
    current team to make its weaknesses and
    resistances more balanced.

    This function first determines which types the
    team is most weak to and which types the team is
    most resistant to via `calc_resistances` and
    `calc_weaknesses`, and then makes its recommendation
    based on this information.
    In particular, it uses `calc_balance` together a brute-force search of
    all ~700 pokémon to determine its recommendation based on the objective of
    maximizing balance.

    Parameters
    ----------
    current_team : list of strings
        list of up to 5 pokémon names
    n_recommendations : integer
        number of pokemon to recommend (default = 1).
    include_legendaries : boolean
        whether or not to include legendary pokémon in
        the recommendations (default = False).
    include_megas : boolean
        whether or not to include Mega pokémon in
        the recommendations (default = False).
    verbose : boolean
        whether or not to print progress updates during the brute-force search.
    early_stop : boolean
        whether or not to stop the brute force search early
        (for speeding up unit testing) (default = False).

    Returns
    -------
    reccomendation : string
        the name of a pokémon that could be added to the input
        team to make its weaknesses and resistances more balanced.

    Example
    --------
    >>> recommend(['Pikachu', 'Eevee', 'Charizard', ...])
    "Lucario"
    """
    # Handle user input error in case of single-pokemon team
    if isinstance(current_team, str):
        current_team = [current_team]

    # Check user input
    try:
        assert isinstance(
            current_team, list
        ), f'{"current_team should be a list of pokemon names."}'
        assert (
            len(current_team) > 0
        ), f'{"current_team should be a non-empty list of pokemon names."}'
        assert (
            len(current_team) < 6
        ), f'{"current_team should have less than six pokemon names."}'
        assert isinstance(
            current_team[0], str
        ), f'{"current_team should be a list of pokemon names."}'
        assert isinstance(
            n_recommendations, int
        ), f'{"n_recommendations should be an integer."}'
    except AssertionError as ex:
        print(f"Invalid input: {ex}")
        return None

    url = (
        "https://raw.githubusercontent.com/"
        + "UBC-MDS/pokehelpyer/main/data/pokemon.csv"
    )
    pokemon_df = pd.read_csv(url)
    if not include_legendaries:
        pokemon_df = pokemon_df.query("Legendary == False")
    if not include_megas:
        pokemon_df = pokemon_df[~pokemon_df["Name"].str.contains("Mega")]

    team_types = get_types(current_team)
    current_resistances = calc_resistances(team_types)
    current_weaknesses = calc_weaknesses(team_types)

    new_balance_dict = dict()

    # Loop through all posible pokemon that could be added to the team
    for i in range(len(pokemon_df)):
        pkmn = pokemon_df.iloc[i, :]
        pkmn_name = pkmn["Name"]
        pkmn_types = [pkmn["Type 1"], pkmn["Type 2"]]
        if pd.isna(pkmn_types[1]):
            pkmn_types.pop()

        pkmn_resistances = calc_resistances([pkmn_types])
        pkmn_weaknesses = calc_weaknesses([pkmn_types])

        # add the pokemon's resistances to the current team's resistances
        new_resistances = dict()
        for type_combo in current_resistances.keys():
            new_resistances[type_combo] = (
                current_resistances[type_combo]
                + pkmn_resistances[type_combo]
            )

        # add the new pokemon's weaknesses to the current team's weaknesses
        new_weaknesses = dict()
        for type_combo in current_weaknesses.keys():
            new_weaknesses[type_combo] = (
                current_weaknesses[type_combo]
                + pkmn_weaknesses[type_combo]
            )

        new_balance = calc_balance(new_resistances, new_weaknesses)
        new_balance_dict[pkmn_name] = new_balance

        if verbose and i % 100 == 99 or i == 0:
            print(f"Iteration number {i + 1} of {len(pokemon_df)}.")

        if early_stop and i > 30:
            print("Stopping early because `early_stop = True`.")
            print("Normally this is only used for testing.")
            break

    new_balance_df = (
        pd.DataFrame(new_balance_dict, index=["balance"])
        .T.reset_index()
        .rename(columns={"index": "Name"})
        .set_index("Name")
    )

    results_df = new_balance_df.join(
        pokemon_df.set_index("Name"), on="Name"
    ).sort_values(by=["balance", "Total"], ascending=False)

    if n_recommendations == 1:
        return results_df.iloc[0, :].name

    temp_df = results_df
    recommendations = []
    for i in range(n_recommendations):
        recommendations.append(temp_df.iloc[0, :].name)
        current_best_balance = temp_df.iloc[0, :]["balance"]
        # Only to pass style check
        if current_best_balance:
            pass
        temp_df = temp_df.query(
            "balance != @current_best_balance"
        ).sort_values(by=["balance", "Total"], ascending=False)

    return recommendations


def calc_balance(resistances, weaknesses):
    """
    Calculate a measure of how balanced a team is using its
    weaknesses and resistances.

    Higher values indicate a more balanced team.

    Parameters
    ----------
    resistances : dictionary
        obtained from calc_resistances

    weaknesses : dictionary
        obtained from calc_weaknesses

    Returns
    -------
    balance : float
        measure of how balanced the team is.

    Examples
    --------
    >>> bad_team = ['Abomasnow', 'Ferrothorn', 'Parasect']
        # All are doubly weak to fire
    >>> resistances = calc_resistances(get_types(bad_team))
    >>> weaknesses = calc_weaknesses(get_types(bad_team))
    >>> calc_balance(resistances, weaknesses)
    -3.11243

    >>> good_team = ['Spiritomb', 'Garchomp', 'Lucario']
    >>> resistances = calc_resistances(get_types(good_team))
    >>> weaknesses = calc_weaknesses(get_types(good_team))
    >>> calc_balance(resistances, weaknesses)
    25.06687
    """
    # Check user input
    try:
        assert isinstance(
            resistances, dict
        ), r"""Input 1 should be a dictionary of resistance values
            of the form {pkmn_type: <int or float>}."""
        assert isinstance(
            weaknesses, dict
        ), r"""Input 2 should be a dictionary of weakness values
            of the form {pkmn_type: <int or float>}."""
        url = (
            "https://raw.githubusercontent.com/"
            + "UBC-MDS/pokehelpyer/main/data/type_chart.csv"
        )
        types_set = set(
            pd.read_csv(url)["Attacking"].tolist()
        )
        assert (
            set(resistances.keys()) == types_set
        ), """Input 1 should be a dictionary of resistance values obtained
            via `calc_resistances`.\n
            There should be one key for each of the 18 pokemon types
            (including Fairy)."""
        assert (
            set(weaknesses.keys()) == types_set
        ), """Input 2 should be a dictionary of weakness values obtained
            via `calc_weaknesses`.\n
            There should be one key for each of the 18 pokemon types
            (including Fairy)."""
        assert isinstance(
            resistances["Normal"], (float, int)
        ), """Input 1 should be a dictionary of resistance values obtained
            via `calc_resistances`.\n
            The values should be floats or integers."""
        assert isinstance(
            weaknesses["Normal"], (float, int)
        ), """Input 2 should be a dictionary of weakness values obtained
            via `calc_weaknesses`.\n
            The values should be floats or integers."""
    except AssertionError as ex:
        print(f"Invalid input: {ex}")
        return None

    type_advantages = dict()
    for type_combo in resistances.keys():
        delta = resistances[type_combo] - weaknesses[type_combo]

        # Peicewise function to penalize negative values more
        # (i.e. to favor penalizing weaknesses over rewaring resistances)
        if delta >= 0:
            type_advantages[type_combo] = delta ** (3 / 4)
        else:
            type_advantages[type_combo] = -((-delta) ** (3 / 2))

    balance = sum(type_advantages.values())
    return balance
